### PR TITLE
Ex2

### PR DESCRIPTION
**Pull request by YairAmar**
All done, ready for CR

### CODE REVIEW

File: ex2/ex2.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+from utils import load_data, plot_data, train_test_split
+import sys
+from logistic_regression import LogisticRegression
+from sklearn.preprocessing import PolynomialFeatures
+import matplotlib.pyplot as plt
+
+
+def plot_w_regularization(x, y):
```
Comment (yarinshaked) [inline ex2/ex2.py:8]: type hints

File: ex2/ex2.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    x1, y1 = load_data(sys.argv[1], add_bias=False)
+    plot_data(x1, y1)
+    plt.show()
+    poly = PolynomialFeatures(1)
+    x1 = poly.fit_transform(x1)
+    x1_train, y1_train, x1_test, y1_test = train_test_split(x1, y1, ratio=0.8)
+
+    log_reg1 = LogisticRegression(x1_train.shape[1], deg=1)
+    log_reg1.fit(x1_train, y1_train, max_iter=400, llambda=0.)
+    print("the train accuracy of the linear model is: ", log_reg1.accuracy(x1_train, y1_train))
```
Comment (yarinshaked) [inline ex2/ex2.py:62]: use fstring formatting

File: ex2/ex2.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    print("the test accuracy of the linear model is: ", log_reg1.accuracy(x1_test, y1_test))
```
Comment (yarinshaked) [inline ex2/ex2.py:63]: also fstring

File: ex2/logistic_regression.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            y: target
+            max_iter: maximum number of iterations (default 400)
+            llambda: lambda of the regularization argument (default 0.)
+            method: optimization method (default "Nelder-Mead")
+                    for non-linear polynomials use "BFGS"
+        Returns:
+            theta: weights of the logistic regression after training
+            cost_fun: value of the cost function after training
+        """
+        res = optimize.minimize(cost_function, x0=self.theta, args=(x, y, llambda), method=method,
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:36]: res is not a good name, maybe result?

File: ex2/logistic_regression.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        z_vals = z_vals.T
+        cont = plt.contour(x1_grid, x2_grid, z_vals, [0])
+        fmt = {0: 'Lambda = %d' % llambda}
+        plt.clabel(cont, inline=1, fontsize=15, fmt=fmt)
+        plt.title("Decision Boundary")
+        plot_data(x[:, 1:], y)
+
+    def predict(self, x: np.array) -> np.array:
+        """
+        return a prediction over the input data, as a class
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:72]: Caps

File: ex2/logistic_regression.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        Args:
+            x: input data
+
+        Returns:
+            hypo: hypothesis prediction of the data's class
+        """
+        hypo = h(self.theta, x) >= 0.5
+        return hypo
+
+    def accuracy(self, x: np.array, y: np.array) -> float:
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:83]: bad name, maybe compute_accracy?

File: ex2/logistic_regression.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        return a prediction over the input data, as a class
+
+        Args:
+            x: input data
+
+        Returns:
+            hypo: hypothesis prediction of the data's class
+        """
+        hypo = h(self.theta, x) >= 0.5
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:80]: predictions might be a better name

File: ex2/logistic_regression.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        calculates the accuracy of the classifier
+
+        Args:
+            x: input data
+            y: target
+
+        Returns:
+            acc: accuracy, correct classification rate in range 0-1
+        """
+        pred = np.array(self.predict(x)) - y
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:94]: numpy array conversion is not necessary

File: ex2/utils.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
+import numpy as np
+import matplotlib.pyplot as plt
```
Comment (yarinshaked) [inline ex2/utils.py:2]: Alphabetic order

File: ex2/utils.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    plt.ylabel("Exam 2 score")
+    plt.legend(["Admitted", "Not admitted"])
+
+
+def sigmoid(z: np.array) -> np.array:
+    """
+    Applies a sigmoid function over the input
+
+    Args:
+        z: input array. in the Logistic Regression context should be  theta.T @ x
```
Comment (yarinshaked) [inline ex2/utils.py:46]: already hinted at the signature

File: ex2/utils.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    Returns:
+        h: sigmoid function applied over z
+    """
+    sig = 1 / (1 + np.exp(-z))
+    return sig
+
+
+def train_test_split(x: np.array, y: np.array, ratio: float = 0.8) -> tuple:
+    """
+    the
```
Comment (yarinshaked) [inline ex2/utils.py:57]: hahaha

File: ex2/utils.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        y: data labels
+        ratio: ration between train and test (default 0.8 - 80% train)
+
+    Returns:
+        x_train: train data
+        y_train: train labels
+        x_test: test data
+        y_test: test labels
+    """
+    data_len = x.shape[0]
```
Comment (yarinshaked) [inline ex2/utils.py:69]: be consistent, m or data_len

File: ex2/utils.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    indices = np.random.permutation(data_len)
+    train_idx, test_idx = indices[:train_len], indices[train_len:]
+    x_train, x_test = x[train_idx, :], x[test_idx, :]
+    y_train, y_test = y[train_idx, :], y[test_idx, :]
+    return x_train, y_train, x_test, y_test
+
+
+def plot_cost(cost_list: list):
+    """
+
```
Comment (yarinshaked) [inline ex2/utils.py:80]: complete

File: ex2/utils.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    Returns:
+
+    """
+    plt.plot(cost_list)
+    plt.xlabel("iterations")
+    plt.ylabel("cost function")
+    plt.show()
+
+
+def h(theta: np.array, x: np.array) -> np.array:
```
Comment (yarinshaked) [inline ex2/utils.py:93]: bad nameeeeeeeeeeeeee

File: ex2/utils.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    Returns:
+        h: sigmoid function applied over z=x@theta
+    """
+    hypothesis = sigmoid(x @ theta.T)
+    return hypothesis
+
+
+def cost_function(theta: np.array, x: np.array, y: np.array, llambda: float = 0.) -> float:
+    """
+    calculates the cost function with given regularization argument
```
Comment (yarinshaked) [inline ex2/utils.py:109]: caps

File: ex2/utils.py [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        x: input data
+        theta: logistic regression's parameters
+    Returns:
+        h: sigmoid function applied over z=x@theta
+    """
+    hypothesis = sigmoid(x @ theta.T)
+    return hypothesis
+
+
+def cost_function(theta: np.array, x: np.array, y: np.array, llambda: float = 0.) -> float:
```
Comment (yarinshaked) [inline ex2/utils.py:107]: maybe compute_cost_function

**Review by yarinshaked — CHANGES_REQUESTED** [review]

File: ex2/utils.py (hunk 9 above)
Comment (YairAmar) [reply ex2/utils.py:2]: tbh i prefer it this way - i always put higher the more often-imported packages like numpy, then matplotlib, pandas etc.

File: ex2/utils.py (hunk 10 above)
Comment (YairAmar) [reply ex2/utils.py:46]: what do you mean? should i remove args?

File: ex2/utils.py (hunk 11 above)
Comment (YairAmar) [reply ex2/utils.py:57]: :( 
ikr

File: ex1/utils.py [hunk 17]
```diff
@@ -30,8 +30,10 @@ def feature_normalize(x: np.array):
     Args:
         x: input data
     """
-    def normalize(vec): return (vec - np.mean(vec)) / np.std(vec)
-    for i in np.arange(x.shape[const.ONE] - const.ONE) + const.ONE:
+    def normalize(vec):
+        return (vec - np.mean(vec)) / np.std(vec)
+
+    for i in np.arange(x.shape[1] - 1) + 1:
```
Comment (yarinshaked) [inline ex1/utils.py:36]: you can use range(1, x.shape[1])

File: ex2/ex2.py [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+def plot_w_regularization(x: np.ndarray, y: np.ndarray):
+    """
+    Plots 4 graphs of the logistic regression model with polynomial features of degree 6
+    and different values of regularization parameter in the cost function
+
+    Args:
+        x: input data
+        y: target
+    """
+    poly_reg = PolynomialFeatures(6)
```
Comment (yarinshaked) [inline ex2/ex2.py:18]: Single Responsibility! this function should only get the values it need for the plot and plot them

File: ex2/ex2.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    x = poly_reg.fit_transform(x)
+    x_train, y_train, x_test, y_test = train_test_split(x, y)
+    log_reg = LogisticRegression(x_train.shape[1], deg=6)
+
+    llambdas = [0., 1., 10., 100.]
```
Comment (yarinshaked) [inline ex2/ex2.py:23]: name is not clear

File: ex2/ex2.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    log_reg1.fit(x1_train, y1_train, max_iter=400, llambda=0.)
+    train_accuracy = log_reg1.compute_accuracy(x1_train, y1_train)
+    test_accuracy = log_reg1.compute_accuracy(x1_test, y1_test)
+    print(f"the train accuracy of the linear model is: {train_accuracy}")
+    print(f"the test accuracy of the linear model is: {test_accuracy}")
+
+    log_reg1.plot_decision_bounds(x1, y1)
+    plt.show()
+    x2, y2 = load_data(sys.argv[2], add_bias=False)
+    plt.grid()
```
Comment (yarinshaked) [inline ex2/ex2.py:56]: same

File: ex2/ex2.py [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        print(f"for lambda = {llambda}, test accuracy = {test_accuracy}")
+
+    plt.show()
+
+
+def main():
+    x1, y1 = load_data(sys.argv[1], add_bias=False)
+    plot_data(x1, y1)
+    plt.show()
+    poly = PolynomialFeatures(1)
```
Comment (yarinshaked) [inline ex2/ex2.py:42]: extract to preprocess function

File: ex2/logistic_regression.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from utils import compute_hypothesis, compute_cost_function, plot_data
+from scipy import optimize
+from sklearn.preprocessing import PolynomialFeatures
+
+
+class LogisticRegression:
+
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:9]: document class purpose

File: ex2/logistic_regression.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+    def __init__(self, n_features: int, deg: int):
+        """
+        The constructor of the LogisticRegression object
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:12]: dot at EOL

File: ex2/logistic_regression.py [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+        Args:
+            n_features: number of features in the input data
+        """
+        self.poly = PolynomialFeatures(deg)
+        self.theta = np.zeros((n_features, 1))
+
+    def fit(self, x: np.ndarray, y: np.ndarray, max_iter: int = 400, llambda: float = 0.,
+            method: str = "Nelder-Mead") -> tuple:
+        """
+        Trains the logistic regression model
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:23]: same

File: ex2/logistic_regression.py [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+class LogisticRegression:
+
+    def __init__(self, n_features: int, deg: int):
+        """
+        The constructor of the LogisticRegression object
+
+        Args:
+            n_features: number of features in the input data
+        """
+        self.poly = PolynomialFeatures(deg)
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:17]: you can send it as an argument, it was created at the main function

File: ex2/logistic_regression.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        Returns a prediction over the input data, as a class
+
+        Args:
+            x: input data
+
+        Returns:
+            predictions: prediction of the data's class
+        """
+        predictions = compute_hypothesis(self.theta, x) >= 0.5
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:80]: set as const (in general should be inside a .yml configuration file but const if enough for now(: )

File: ex2/logistic_regression.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+        return predictions
+
+    def compute_accuracy(self, x: np.ndarray, y: np.ndarray) -> float:
+        """
+        Calculates the accuracy of the classifier
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:85]: same

File: ex2/logistic_regression.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+
+        Args:
+            x: input data
+            y: target
+
+        Returns:
+            acc: accuracy, correct classification rate in range 0-1
+        """
+        pred = self.predict(x) - y
```
Comment (yarinshaked) [inline ex2/logistic_regression.py:94]: use == instead

File: ex2/utils.py [hunk 29]
```diff
@@ -0,0 +1,128 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+def load_data(path: str, add_bias: bool = True) -> tuple:
+    """
+    Loading data and formatting for the latter linear regression
```
Comment (yarinshaked) [inline ex2/utils.py:7]: same

File: ex2/utils.py (hunk 10 above)
Comment (yarinshaked) [reply ex2/utils.py:46]: you already hinted

File: ex2/utils.py [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+    Returns:
+        h: sigmoid function applied over z
+    """
+    sig = 1 / (1 + np.exp(-z))
+    return sig
+
+
+def train_test_split(x: np.ndarray, y: np.ndarray, ratio: float = 0.8) -> tuple:
+    """
+    Splits the data into 2 groups - train and test, in a proportion given by ratio
```
Comment (yarinshaked) [inline ex2/utils.py:57]: same

File: ex2/utils.py [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+    plt.scatter(x1[y[:, 0] == 1], x2[y[:, 0] == 1], marker='+')
+    plt.scatter(x1[y[:, 0] == 0], x2[y[:, 0] == 0], marker='3')
+    plt.xlabel("Exam 1 score")
+    plt.ylabel("Exam 2 score")
+    plt.legend(["Admitted", "Not admitted"])
+
+
+def sigmoid(z: np.ndarray) -> np.ndarray:
+    """
+    Applies a sigmoid function over the input
```
Comment (yarinshaked) [inline ex2/utils.py:43]: same

File: ex2/utils.py [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
+    indices = np.random.permutation(m)
+    train_idx, test_idx = indices[:train_len], indices[train_len:]
+    x_train, x_test = x[train_idx, :], x[test_idx, :]
+    y_train, y_test = y[train_idx, :], y[test_idx, :]
+    return x_train, y_train, x_test, y_test
+
+
+def plot_cost(cost_list: list):
+    """
+    Plots all values of the cost-function acquired through the training of the model
```
Comment (yarinshaked) [inline ex2/utils.py:81]: same

File: ex2/utils.py [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    plt.plot(cost_list)
+    plt.xlabel("iterations")
+    plt.ylabel("cost function")
+    plt.show()
+
+
+def compute_hypothesis(theta: np.ndarray, x: np.ndarray) -> np.ndarray:
+    """
+    Computes the hypothesis function
```
Comment (yarinshaked) [inline ex2/utils.py:97]: same

File: ex2/utils.py [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+    Returns:
+        h: sigmoid function applied over z=x@theta
+    """
+    hypothesis = sigmoid(x @ theta.T)
+    return hypothesis
+
+
+def compute_cost_function(theta: np.ndarray, x: np.ndarray, y: np.ndarray, llambda: float = 0.) -> float:
+    """
+    Calculates the cost function with given regularization argument
```
Comment (yarinshaked) [inline ex2/utils.py:111]: same

**Review by yarinshaked — CHANGES_REQUESTED** [review]

File: ex2/utils.py (hunk 10 above)
Comment (YairAmar) [reply ex2/utils.py:46]: fixed

File: ex1/utils.py (hunk 17 above)
Comment (YairAmar) [reply ex1/utils.py:36]: done

File: ex2/ex2.py (hunk 19 above)
Comment (YairAmar) [reply ex2/ex2.py:23]: fixed -> lambda_values

File: ex2/ex2.py (hunk 20 above)
Comment (YairAmar) [reply ex2/ex2.py:56]: done

File: ex2/ex2.py (hunk 21 above)
Comment (YairAmar) [reply ex2/ex2.py:42]: done

File: ex2/utils.py (hunk 30 above)
Comment (YairAmar) [reply ex2/utils.py:57]: done

File: ex2/logistic_regression.py (hunk 22 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:9]: done

File: ex2/logistic_regression.py (hunk 23 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:12]: done

File: ex2/logistic_regression.py (hunk 24 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:23]: done

File: ex2/logistic_regression.py (hunk 26 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:80]: named the constant THRESHOLD

File: ex2/logistic_regression.py (hunk 27 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:85]: done

File: ex2/logistic_regression.py (hunk 28 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:94]: done

File: ex2/utils.py (hunk 29 above)
Comment (YairAmar) [reply ex2/utils.py:7]: fixed

File: ex2/utils.py (hunk 32 above)
Comment (YairAmar) [reply ex2/utils.py:81]: fixed

File: ex2/utils.py (hunk 33 above)
Comment (YairAmar) [reply ex2/utils.py:97]: fixed

File: ex2/utils.py (hunk 31 above)
Comment (YairAmar) [reply ex2/utils.py:43]: done

File: ex2/logistic_regression.py (hunk 25 above)
Comment (YairAmar) [reply ex2/logistic_regression.py:17]: cant do since PolynomialFeatures is now used from within the preprocessing method

File: ex2/ex2.py (hunk 18 above)
Comment (YairAmar) [reply ex2/ex2.py:18]: have to train the models in order to plot the decision boundaries anyways.

**Pull request merged**